### PR TITLE
e2e tests: libvmi: add annotation with test name to new VMIs

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
This PR adds an annotation to VMIs created by libvmi with the name of the current test (if available).
This will help identify which test created any rogue VMI.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/libvmi/vmi.go [hunk 1]
```diff
@@ -20,6 +20,7 @@
 package libvmi
 
 import (
+	"github.com/onsi/ginkgo/v2"
```
Comment (dankenigsberg) [inline tests/libvmi/vmi.go:23]: I think it's a great idea to annotate all test VMs, but I don't think that it is acceptable for `libvmi` to depend on the framework that uses it. Can we find a way to inject this annotation without adding this dependency?

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/libvmi/vmi.go (hunk 1 above)
Comment (jean-edouard) [reply tests/libvmi/vmi.go]: Agreed. Pushed v2 without the dependency, PTAL!

Comment (EdDev) [reply tests/libvmi/vmi.go]: How about we create infra to register default options by users at `init()`?

Comment (EdDev) [reply tests/libvmi/vmi.go]:
```go
func init() {
  libvmi.RegisterDefaultOptions(withTestAnnotation)
}
```

Comment (jean-edouard) [reply tests/libvmi/vmi.go]: > ```go
> func init() {
>   libvmi.RegisterDefaultOption(withTestAnnotation)
> }
> ```

Pushed something like that, PTAL

File: tests/libvmi/factory.go [hunk 2]
```diff
@@ -88,7 +89,7 @@ func NewAlpineWithTestTooling(opts ...Option) *kvirtv1.VirtualMachineInstance {
 }
 
 func cirrosMemory() string {
-	if checks.IsARM64(testsuite.Arch) {
+	if checks.IsARM64(libnode.GetArch()) {
```
Comment (dankenigsberg) [inline tests/libvmi/factory.go:92]: oh, `libvmi` already had a dependency on `testsuite`. I think that was *my* fault, long ago. Depending on `libnode` is better, but not great. I wonder if (on another PR) we can have the caller supply the architecture somehow?

```go
func init() {
  libvmi.RegisterDefaultOption(withTestAnnotation)
  libvmi.SetArch(libnode.GetArch)
}
```

File: tests/vmipreset_test.go [hunk 3]
```diff
@@ -380,8 +380,6 @@ var _ = Describe("[rfe_id:609][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 			Expect(err).ToNot(HaveOccurred())
 
 			By("Verifying VMI")
-			Expect(newVmi.Annotations).To(BeNil())
```
Comment (dankenigsberg) [inline tests/vmipreset_test.go:383]: isn't this one important ? (I don't know enough about presets to tell)
maybe this should be replaced by checking that we only have the testsuite's annotation?

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/vmipreset_test.go (hunk 3 above)
Comment (EdDev) [reply tests/vmipreset_test.go:383]: Now that there is a default annotation always added, I guess this assertion is not correct anymore.
But it makes sense to ask if the intention here was to check that no one else is adding a new annotation. Perhaps this assert should be changed with checking the original annotation from before the creation and the one after. (i.e. `Expect(vmi.Annotations).To(Equal(newVmi.Annotations))`)

File: tests/libvmi/vmi.go [hunk 4]
```diff
@@ -46,6 +46,12 @@ func New(opts ...Option) *v1.VirtualMachineInstance {
 	return vmi
 }
 
+var defaultOption Option
+
+func RegisterDefaultOption(opt Option) {
+	defaultOption = opt
+}
```
Comment (EdDev) [inline tests/libvmi/vmi.go:53]: Very nice, thanks!

Can we extend this to support many such options?

```go
var defaultOptions []Option

func RegisterDefaultOption(opt Option) {
	defaultOptions = append(defaultOptions, opts)
}
```

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by xpivarc** [comment]
Note: There is a considerable amount of tests that do not use libvmi.

Note 2:  I think this would be better to have as part of client/framework but that would require more time to make it work.

**Comment by EdDev** [comment]
> Note: There is a considerable amount of tests that do not use libvmi.
> 
> Note 2: I think this would be better to have as part of client/framework but that would require more time to make it work.

Tests that want to take advantage of this marking need to use libvmi, I think it is a reasonable requirement.

I do not think there is a need to delay this change because it does not apply to all VMIs created by the tests.

**Review by dankenigsberg — APPROVED** [review]

**Review by EdDev — APPROVED** [review]

**Comment by EdDev** [comment]
> please ask for approval from [jean-edouard](https://github.com/jean-edouard).

The bot recommends the author for approval :)

**Review by dhiller — COMMENTED** [review]
This is a great help for when things go south, thank you @jean-edouard !

With a bit of advertisement we could even convince people to use `libvmi` more.

/approve

**Review by brianmcarey — COMMENTED** [review]
/hold

@jean-edouard - its looks like there is a test that keeps failing on the sig-compute lanes here. 
`[sig-compute]VMIPreset Override [test_id:644][rfe_id:609] should override presets`

Placing a hold just to stop it needlessly retesting.

**Comment by jean-edouard** [comment]
> /hold
> 
> @jean-edouard - its looks like there is a test that keeps failing on the sig-compute lanes here. `[sig-compute]VMIPreset Override [test_id:644][rfe_id:609] should override presets`
> 
> Placing a hold just to stop it needlessly retesting.

Sorry about that and thank you for holding! I accidentally removed the fix with the last cleanup. Should be good now.
/hold cancel

**Review by dankenigsberg — APPROVED** [review]

**Pull request merged**